ir/string.go

```diff
@@ -8,17 +8,24 @@ import (
 
 // String is a node representing a string literal.
 type String struct {
-	Value string
+	Value       string
+	metadataCtx *metadataContext
 }
 
 func (s String) Disassemble() xr.Node {
 	return xr.String{Value: s.Value}
 }
 
+func (s String) Metadata() MetadataInfo {
+	return s.metadataCtx.getMetadata()
+}
+
 func (s String) UpdateWith(ctx UpdateContext, with Node) (Node, error) {
 	w, ok := with.(String)
 	if !ok {
 		return nil, fmt.Errorf("cannot update with a non-string")
 	}
+	// Update metadata
+	s.metadataCtx.update(w.metadataCtx)
```

Comment:
It looks like you are updating the metadata of s, but you are returning w as a result.
Since IR nodes are mutable (unlike XR nodes), the right solution might be to
make all IR methods mutable, e.g. (s *String), and change the interface of UpdateWith to:

(s *String) UpdateWith(ctx UpdateContext, with Node) error

Reply:
Good call. This was a typo I missed completely in my review. I actually wanted IR to be immutable as XR. If you check the code of `UpdateWith` for `Dict`or `Set` you'll see that there is done right, I return a copy of the input with the updates.

I fixed it keeping IR nodes immutable. Let me know if you think it would be beter to make IR nodes immutable, and I can make the change.

Reply:
Thats up to you. This code is internal to the VM. There is not a very good reason to keep them as values, because:
- this may end up confusing you into making subtle bugs in the future
- it is slightly less efficient
If you want to keep it as is, maybe add a test that makes sure that the bug I caught does not slip in again when people add new types on meta fields.

Reply:
You are right, makes sense. Will update it tomorrow before merging.